tracer/propagator.go

```diff
@@ -0,0 +1,52 @@
+package tracer
+
+import (
+	"errors"
+
+	"github.com/DataDog/dd-trace-go/dd"
+)
+
+// Propagator implementations should be able to inject and extract
+// SpanContexts into an implementation specific carrier.
+type Propagator interface {
+	// Inject takes the SpanContext and injects it into the carrier using
+	// an implementation specific method.
+	Inject(context dd.SpanContext, carrier interface{}) error
+
+	// Extract returns the spanContext from the given carrier using an
+	// implementation specific method.
+	Extract(carrier interface{}) (dd.SpanContext, error)
+}
+
+// TextMapWriter allows setting key/value pairs of strings on the underlying
+// data structure.
+type TextMapWriter interface {
+	// Set sets the given key/value pair.
+	Set(key, val string)
+}
+
+// TextMapReader allows iterating over sets of key/value pairs.
+type TextMapReader interface {
+	// ForeachKey iterates over all keys that exist in the underlying
+	// implementation. It takes a callback function which should be called
+	// using all key/value pairs as arguments. On the first error returned,
+	// ForeachKey should return that error.
+	ForeachKey(handler func(key, val string) error) error
+}
+
```

Comment:
Wonder if we need a:
```go
type Carrier interface {
     TextMapWriter
     TextMapReader
}
```
It looks like most real-world implementations implement both writer & reader anyway. Still, I'm totally in having separate interfaces for reader & writer. Again, I let you decide in the end, was just wondering if it could be worth it.

Reply:
I was wondering pondering on this too. I was thinking to add your suggestion and to change the `carrier` param of `Inject` and `Extract` to use that instead of `interface{}`. I've decided against it because the carrier can also be of a binary type in the future, not just a map. In that case we'll have to cast accordingly. For example, the OpenTracing interfaces encourage all implementations to support binary as well as text map.